presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/Lookup.java

```diff
@@ -35,15 +34,17 @@
     @Deprecated
     default PlanNode resolve(PlanNode node)
     {
-        return getOnlyElement(resolveGroup(node));
+        if (node instanceof GroupReference) {
+            return getOnlyElement(resolveGroup(node));
+        }
+        return node;
     }
 
     /**
      * Resolves nodes by materializing GroupReference nodes
      * representing symbolic references to other nodes.
      * <p>
-     * If the node is not a GroupReference, it returns the
-     * singleton of the argument node.
+     * @throws IllegalArgumentException if the node is not a GroupReference
```

Comment:
If we plan to make it strict, then I would prefer to also have a strict type check and make `resolveGroup` accept a `GroupReference` instead of suggesting that it works with other type of nodes (especially when semantics of `resolve` is different).

Reply:
I'm against, as this would move the burden of casting to callers, making rule code that needs to use `resolveGroup` even less readable.

Reply:
> needs to use resolveGroup even less readable
but less error prone. I think it will be pattern matching that will mostly use it anyway

Reply:
IMO It won't be less err-prone, since there is no help form type-system here. So end up casting whenever needed.

Reply:
I would just leave with the most basic method:
```
List<PlanNode> resolveGroup(GroupReference node);
```
If we find that throwing `List<PlanNode> resolveGroup(PlanNode node);` is more useful then we can change it later on, but at this point there is no need for upcast in `resolveGroup`. I think it won't be used except for pattern matching and maybe join enumeration.

Reply:
I think both approaches are fine. It should not matter much as long as this method is seldomly used.

Reply:
i agree with @findepi